src/sdam/description/topology/server_selection/mod.rs

```diff
@@ -1,25 +1,91 @@
 #[cfg(test)]
 mod test;
 
-use std::{fmt, time::Duration};
+use std::{collections::HashMap, fmt, ops::Deref, sync::Arc, time::Duration};
 
-use rand::seq::IteratorRandom;
+use rand::{
+    rngs::SmallRng,
+    seq::{IteratorRandom, SliceRandom},
+    SeedableRng,
+};
 
 use super::TopologyDescription;
 use crate::{
     error::{ErrorKind, Result},
+    options::StreamAddress,
     sdam::{
         description::{
             server::{ServerDescription, ServerType},
             topology::TopologyType,
         },
         public::ServerInfo,
+        Server,
     },
     selection_criteria::{ReadPreference, SelectionCriteria, TagSet},
 };
 
 const DEFAULT_LOCAL_THRESHOLD: Duration = Duration::from_millis(15);
 
+/// Struct encapsulating a selected server that handles the opcount accounting.
+#[derive(Debug)]
+pub(crate) struct SelectedServer {
+    server: Arc<Server>,
+}
+
+impl SelectedServer {
+    fn new(server: Arc<Server>) -> Self {
+        server.increment_operation_count();
+        Self { server }
+    }
+}
+
+impl Deref for SelectedServer {
+    type Target = Server;
+
+    fn deref(&self) -> &Server {
+        self.server.deref()
+    }
+}
+
+impl Drop for SelectedServer {
+    fn drop(&mut self) {
+        self.server.decrement_operation_count();
+    }
+}
+
+/// Attempt to select a server, returning None if no server could be selected
+/// that matched the provided criteria.
+pub(crate) fn attempt_to_select_server<'a>(
+    criteria: &'a SelectionCriteria,
+    topology_description: &'a TopologyDescription,
+    servers: &'a HashMap<StreamAddress, Arc<Server>>,
+) -> Result<Option<SelectedServer>> {
+    let in_window = topology_description.suitable_servers_in_latency_window(criteria)?;
+    let in_window_servers = in_window
+        .into_iter()
+        .flat_map(|desc| servers.get(&desc.address))
+        .collect();
+    Ok(select_server_in_latency_window(in_window_servers).map(SelectedServer::new))
+}
+
+/// Choose a server from several suitable choices within the latency window according to
+/// the algorithm laid out in the server selection specification.
+fn select_server_in_latency_window(in_window: Vec<&Arc<Server>>) -> Option<Arc<Server>> {
+    if in_window.is_empty() {
+        return None;
+    } else if in_window.len() == 1 {
+        return Some(in_window[0].clone());
+    }
+
+    let mut rng = SmallRng::from_entropy();
+    let mut choices = in_window.into_iter().choose_multiple(&mut rng, 2);
+    choices.shuffle(&mut rng);
```

Comment:
I think the shuffling can be avoided by using calling `choose_multiple` [directly on the slice](https://docs.rs/rand/0.7.3/rand/seq/trait.SliceRandom.html#tymethod.choose_multiple) rather than on the iterator.

Reply:
Ah cool, fixed.